rts/motoko-rts/src/idl.rs

```diff
@@ -486,3 +512,310 @@ unsafe extern "C" fn skip_fields(tb: *mut Buf, buf: *mut Buf, typtbl: *mut *mut
         *n -= 1;
     }
 }
+
+unsafe fn is_null_opt_reserved(typtbl: *mut *mut u8, end: *mut u8, t: i32) -> bool {
+    if is_primitive_type(t) {
+        return t == IDL_PRIM_null || t == IDL_PRIM_reserved;
+    }
+
+    // unfold t
+    let mut t = t;
+
+    let mut tb = Buf {
+        ptr: *typtbl.add(t as usize),
+        end: end,
+    };
+
+    t = sleb128_decode(&mut tb);
+
+    return t == IDL_CON_opt;
+}
+
+// https://github.com/dfinity/candid/blob/master/rust/candid/src/types/subtype.rs#L10
+// https://github.com/dfinity/candid/blob/20b84d1c1515e2c1db353ebe02b738486f835466/spec/Candid.md
+unsafe fn sub(
+    rel: &BitRel,
+    p: bool,
```

Comment:
Is `p` needed because we don't unset the cache when subtyping fails?

Reply:
p tracks the polarity of the subtype test and flips in contravariant positions. It's used to determine which half of the cache to read and the interpretation of the type indices (which also flips).

Reply:
Yep